r-lib-error-1.0.md

```diff
@@ -0,0 +1,60 @@
+
+# Specification of r-lib errors, version 1.0
+
+Note: this is an experimental draft.
+
+Note: version 1.0 corresponds to the `rlang_error` class, as
+implemented in rlang 0.4.0.
+
+## Classes
+
+An r-lib error object always has at least classes `rlib_error`, `error`
+and `condition`. It may have other classes.
+
+## Members
+
+An `rlib_error` object has the following members:
+
+* `message`: `character(1)`, the error message.
+* `trace`: optional, an `rlib_trace` object, see below. 
+* `parent`: optional, pointer to a parent error, of class `error`. It
+  may also be `NULL`, which is the same as missing.
+* `call`: optional, the captured call, of class `call`.
+* `version`: optional, `r_lib` error version, defaults to 1.0, i.e. this
+  specification.
+
+## Traces
+
+An `rlib_trace` object has the following members:
+
+* `calls`: A list of calls, each of class `call`. This is the stack when
+  the error was thrown. `calls[[1]]` is the top of the stack.
+* `parents`: Integer vector of parent frame pointers, as in `sys.parents()`.
+* `envs`: List or character vector of environment names. They are typically
```

Comment:
Should we make it `ids`? They only need to be a character vector of unique identifiers for the environment frames. For simplicity I don't think we should allow list though? Should rlang switch to character vector?

> They are typically package names for frames that correspond to functions in packages.

I don't think this is correct. I wouldn't mention "names" here, an execution frame normally doesn't have a name, it's only the case with expressions manually evaluated with R-level `eval()`.

Reply:
Actually we may not even need these ids with the calling-handler approach to capturing chained errors. We probably should try it first.

Reply:
Right, I mixed this up with the environments of the functions. So these are just the execution environments, I'll fix that. 

Re `ids`, I think for version 1.0 we should basically just write down what rlang currently does.

Reply:
Given the very strict contract offered in this spec, I think we should get things right first.

Reply:
Well, there is already an implementation in rlang, and another one in callr etc. :) So I think maybe we can write down the current rlang state as 1.0 (or we can also call it 0.9 if you like), and then immediately "fix" it as the next version.

Reply:
Maybe I'll remove the backwards compatibility promise then? Or have that promise for the "fixed" 1.0 version?